### PR TITLE
plan: refactor errors

### PR DESCRIPTION
**Pull request by zz-jason**
1. move all error definitions into `errors.go`
2. remove `terror.ClassOptimizerPlan`

### CODE REVIEW

**Comment by zz-jason** [comment]
@winoros @XuHuaiyu PTAL

**Comment by XuHuaiyu** [comment]
/go/src/github.com/pingcap/tidb/plan/errors.go:50:2: exported var ErrUnsupportedType should have comment or be unexported

File: executor/prepared_test.go [hunk 1]
```diff
@@ -260,5 +260,5 @@ func (s *testSuite) TestPreparedNameResolver(c *C) {
 	tk.MustExec("create table t (id int, KEY id (id))")
 	tk.MustExec("prepare stmt from 'select * from t limit ? offset ?'")
 	_, err := tk.Exec("prepare stmt from 'select b from t'")
-	c.Assert(err.Error(), Equals, "[plan:1054]Unknown column 'b' in 'field list'")
+	c.Assert(err.Error(), Equals, "[optimizer:1054]Unknown column 'b' in 'field list'")
```
Comment (winoros) [inline executor/prepared_test.go]: planner？

**Comment by zz-jason** [comment]
@XuHuaiyu @winoros PTAL again

File: plan/logical_plan_builder.go [hunk 2]
```diff
@@ -774,14 +774,14 @@ func (b *planBuilder) buildLimit(src LogicalPlan, limit *ast.Limit) LogicalPlan
 	if limit.Offset != nil {
 		offset, err = getUintForLimitOffset(sc, limit.Offset.GetValue())
 		if err != nil {
-			b.err = ErrWrongArguments
+			b.err = ErrWrongArguments.GenByArgs("LIMIT")
```
[the remark this answers is not in the document]
Comment (zz-jason) [reply plan/logical_plan_builder.go]: the `Offset` and `Count` of `ast.Limit` is guaranteed by parser, maybe we should delete this check.

[the remark this answers is not in the document]
Comment (XuHuaiyu) [reply plan/logical_plan_builder.go]: But if offset and count are paramMarker, they are not guaranteed to be uint64.

**Comment by XuHuaiyu** [comment]
LGTM

**Comment by zz-jason** [comment]
@winoros PTAL

**Comment by zimulala** [comment]
@zz-jason Please fix the confliction

**Comment by zz-jason** [comment]
@zimulala done, PTAL

**Pull request merged**